### PR TITLE
feat(flow): add execution time estimation for flows

### PR DESCRIPTION
**Pull request by ashtarkb**
- Add time_estimator.py utility for predicting full dataset execution time
- Implement estimate_total_time() method in Flow class
- Cache dry_run results for efficient time estimation
- Support both sequential and concurrent time predictions
- Add comprehensive test coverage for time estimation (14 tests)
- Update documentation with time estimation examples and best practices
- Include accuracy disclaimer noting 15-40% over-estimation for planning

This feature helps users predict execution time and estimate total LLM requests before processing large datasets.

## Summary by CodeRabbit

- **New Features**
  - Optional time-estimation during dry-run: predicts total runtime, estimated API/LLM requests, per-block breakdown, supports max concurrency, and shows a Rich summary.

- **Documentation**
  - Time Estimation guidance and examples added to Quick Start, Concepts, and Flows overview (how it works, accuracy, when to use, runtime-param notes).

- **Bug Fixes / Validation**
  - Validation added for max_concurrency (must be a positive integer).

- **Tests**
  - Extensive test coverage for time estimation, concurrency, metrics display, and edge cases.

### CODE REVIEW

**Review by shivchander — CHANGES_REQUESTED** [review]
Thanks for contributing! I think we can achieve the same user value with a much simpler approach.

Instead of a separate estimation system, we could extend the existing `dry_run()` method with an optional flag:

Something like this: 

```python
  def dry_run(
      self,
      dataset: Dataset,
      sample_size: int = 2,
      runtime_params: Optional[dict[str, dict[str, Any]]] = None,
      estimate_full_time: bool = False,  # NEW
      max_concurrency: Optional[int] = None,  # NEW
  ) -> dict[str, Any]:

# At the end of the dry run
  if estimate_full_time:
      scaling_factor = len(dataset) / actual_sample_size
      estimated_time = execution_time * scaling_factor

      # Simple concurrency benefit for LLM blocks
      if max_concurrency and max_concurrency > 1:
          has_llm_blocks = any('LLM' in b['block_type'] for b in dry_run_results["blocks_executed"])
          if has_llm_blocks:
              estimated_time /= min(max_concurrency ** 0.5, 10)  # Diminishing returns

      estimated_time *= 1.25  # 25% buffer

      # Display in table format (like existing metrics display)
      display_time_estimation_table(estimated_time, len(dataset), max_concurrency)

  return dry_run_results 
```

And then we can add the `display_time_estimation_table` similar to what we have in https://github.com/Red-Hat-AI-Innovation-Team/sdg_hub/blob/main/src/sdg_hub/core/utils/flow_metrics.py#L71

This way we could simplify the api to something like this: 

```python
 # Get dry run AND time estimate in one call
  result = flow.dry_run(dataset, sample_size=5, estimate_full_time=True, max_concurrency=50)
```

And then display a table. 


Would you be open to exploring this simpler direction? Happy to help with the implementation if you'd like to try this approach!

**Comment by shivchander** [comment]
Thanks for testing! Ye that seems like a significant difference, I agree that we can use best of both worlds. Here's what I am thinking: 

* We can refactor `estimate_total_time` to a private internal method 
* Expose the control through `dry_run` api

```python
 # Get dry run AND time estimate in one call
 result = flow.dry_run(dataset, sample_size=5, estimate_full_time=True, max_concurrency=50)
 ```

We can retain the same api, but when `estimate_full_time=True` we can call your method and generate with sample size 1 and 5 (or whatever is set). And then return the results for sample size 5 and then log a rich summary table for time estimate. 

What do you think about this

File: src/sdg_hub/core/flow/base.py [hunk 1]
```diff
@@ -1181,6 +1214,13 @@ def dry_run(
                 f"in {execution_time:.2f}s"
             )
 
+            # Perform time estimation if requested
+            if estimate_full_time:
+                estimation = self._estimate_total_time(
+                    dry_run_results, dataset, runtime_params, max_concurrency
+                )
+                dry_run_results["time_estimation"] = estimation
```
Comment (shivchander) [inline src/sdg_hub/core/flow/base.py]: Could we not add the time estimation to the dataset, I think having it in the table we log would suffice

File: docs/flows/overview.md [hunk 2]
```diff
@@ -264,6 +264,44 @@ print(f"Output columns: {dry_result['final_dataset']['columns']}")
...
+### Time Estimation
+
+Predict execution time for your full dataset before running:
+
+```python
+# Get dry run results AND time estimation in one call
+result = flow.dry_run(
+    dataset, 
+    sample_size=5, 
+    estimate_full_time=True,
```
Comment (shivchander) [inline docs/flows/overview.md]: Could we rename this to `enable_time_estimation`, I think that reads less ambiguous

File: docs/flows/overview.md [hunk 3]
```diff
@@ -264,6 +264,44 @@ print(f"Output columns: {dry_result['final_dataset']['columns']}")
...
+    max_concurrency=100
+)
+
+# Time estimation automatically displayed in Rich table format
+# Access estimation data programmatically:
+if 'time_estimation' in result:
```
Comment (shivchander) [inline docs/flows/overview.md]: similar to https://github.com/Red-Hat-AI-Innovation-Team/sdg_hub/pull/447/files#r2403077138, lets not add fields to the dataset, adding to the table should suffice

File: docs/concepts.md [hunk 4]
```diff
@@ -148,9 +148,24 @@ Every block validates data at runtime:
...
 
+```python
+# Test AND estimate in one call
+result = flow.dry_run(dataset, sample_size=5, estimate_full_time=True, max_concurrency=100)
+
+# Access dry run results
+print(f"Tested with {result['sample_size']} samples")
+
+# Access time estimation
+if 'time_estimation' in result:
```
Comment (shivchander) [inline docs/concepts.md]: similar to https://github.com/Red-Hat-AI-Innovation-Team/sdg_hub/pull/447/files#r2403107775

**Review by shivchander — CHANGES_REQUESTED** [review]

**Review by shivchander — APPROVED** [review]
